Client/SafeBrowsing/GSafeBrowsingDatabase.swift

```diff
@@ -0,0 +1,509 @@
+                //hashes.sort(by: { $0.lexicographicallyPrecedes($1) })
+                
+                response.removals.forEach({
+                    $0.rawIndices?.indices.forEach({
+                        if Int($0) > 0 && Int($0) < hashes.count {
+                            hashes[Int($0)] = Data()
+                        }
+                    })
+                })
+                
+                hashes = hashes.filter({ !$0.isEmpty })
+                
+                response.additions.forEach({
+                    guard let rawHashes = $0.rawHashes else {
+                        return
+                    }
+                    
+                    if let data = Data(base64Encoded: rawHashes.rawHashes) {
+                        let strideSize = Int(rawHashes.prefixSize)
+                        
+                        for i in stride(from: 0, to: data.count, by: strideSize) {
+                            let startIndex = data.index(data.startIndex, offsetBy: i)
+                            let endIndex = data.index(startIndex, offsetBy: strideSize)
+                            hashes.append(data.subdata(in: startIndex..<endIndex))
+                        }
+                    }
+                })
+                
+                //Hashes must be sorted
+                hashes.sort(by: { $0.lexicographicallyPrecedes($1) })
+                
+                threat.hashes?.forEach({
+                    if let hash = $0 as? ThreatHash {
+                        backgroundContext.delete(hash)
+                    }
+                })
+                
+                threat.hashes = NSOrderedSet(array: hashes.map({
+                    let hash = ThreatHash(context: backgroundContext)
+                    hash.hashPrefix = $0.rawString()
+                    hash.hashData = $0
+                    return hash
+                }))
+                
+                if !response.checksum.sha256.isEmpty {
+                    threat.checksum = response.checksum.sha256
+                }
+                
+                if !self.validate(response.checksum.sha256, hashes) {
+                    backgroundContext.rollback()
+                    return completion(SafeBrowsingError("Threat List Checksum Mismatch"))
+                }
+                
+                self.saveContext(backgroundContext)
+            }
+        })
+    }
+    
+    func scheduleUpdate(onNeedsUpdating: @escaping () -> Void) {
+        updateTimer?.invalidate()
+        updateTimer = Timer.scheduledTimer(withTimeInterval: self.lastFetchWaitDuration, repeats: false) {
+            $0.invalidate()
+            DispatchQueue.main.async {
+                onNeedsUpdating()
+            }
+        }
+    }
+    
+    private func calculateBackoffTime(_ numberOfRetries: Int16) -> Double {
+        let minutes = Double(1 << Int(numberOfRetries - 1)) * (15.0 * 60.0) * Double.random(in: 0...1) + 1
+        return min(minutes, 24 * 60)
```

Comment:
It does seem strange that we're returning minutes here where anywhere else we are dealing with seconds. Maybe we should return `min(minutes*60, 24*60*60)` instead?